Fix weird UI when dragging a panel with linker overlay open

jason-wang24 opened the pull request:
Fixes #964 
- Added `PanelEvent.DRAGGING` and `PanelEvent.DROPPED` events and `isDraggingPanel` state to `Linker.tsx`. isDraggingPanel is set to false if the panel being dragged has a link starting or ending in that panel and set back to true when the panel is dropped.
- Perhaps we can also reopen the linker overlay when the panel is dropped (this would require adding a new event but is otherwise trivial)

dsmmcken reviewed (dismissed):
After trying it, I agree. You are right we should hide it then re-show it instead.

Could make it unmount the children, or display none on a parent, but it should be with zero-transition as picking up the component is instant. Looks still a bit weird to have it fade out in the wrong spot as they fade out.

jason-wang24 commented:
I added a `showOverlay` state to `Linker.tsx` which gets toggled by the DRAGGING and DROPPED events. I'm slightly concerned this is not defensive enough, as if the panel is somehow not dropped, `showOverlay` may be stuck being false. Not sure if this is a legitimate concern or not but either way I'm not sure how to make it more defensive (set it to true elsewhere?) if need be.

dsmmcken commented:
react.development.js:1139 Uncaught Error: React.Children.only expected to receive a single React element child.
    at Object.onlyChild [as only] (react.development.js:1139:13)
    at Transition2.render (Transition.js:408:108)
    at finishClassComponent (react-dom.development.js:17485:31)
    at updateClassComponent (react-dom.development.js:17435:24)
  
  Seeing this warning

dsmmcken commented:
(fat fingered the close with comment button, ignore)

mofojed commented on packages/dashboard-core-plugins/src/linker/Linker.tsx:
```diff
@@ -666,7 +702,7 @@ export class Linker extends Component<LinkerProps, LinkerState> {
         : 'Create a link between the source column button and a table column by clicking on one, then the other. Remove the link by clicking it directly. Click done when finished.';
 
-    return (
+    return isLinkOverlayShown ? (
```
Shouldn't need a ternary here, just keep it in the `CSSTransition` `in` attribute - then there will be an appropriate fade animation when you start dragging, and another fade animation when you drop.

jason-wang24 replied:
> After trying it, I agree. You are right we should hide it then re-show it instead.
> 
> Could make it unmount the children, or display none on a parent, but it should be with zero-transition as picking up the component is instant. Looks still a bit weird to have it fade out in the wrong spot as they fade out.

I think we are trying to avoid the fade animation. Is there a better way to get the overlay to disappear as quickly as possible?

mofojed commented on packages/dashboard-core-plugins/src/linker/Linker.tsx:
```diff
@@ -108,9 +109,15 @@ export class Linker extends Component<LinkerProps, LinkerState> {
+    this.state = {
+      linkInProgress: undefined,
+      selectedIds: new Set<string>(),
+      showOverlay: true,
```
I'd name this `isDraggingLinkedPanel` or even just `isDraggingPanel` instead, as it's kind of confusing naming it `showOverlay` when the overlay is shown based on the active tool in `isOverlayShown()` method

mofojed commented on packages/dashboard-core-plugins/src/linker/Linker.tsx:
```diff
@@ -108,9 +109,15 @@ export class Linker extends Component<LinkerProps, LinkerState> {
+    this.state = {
+      linkInProgress: undefined,
+      selectedIds: new Set<string>(),
+      isDraggingPanel: true,
```
I think you've got the `boolean` backwards for the way this is named. `isDraggingPanel` should be `true` while there's a panel that's currently being dragged.

mofojed commented on packages/dashboard-core-plugins/src/linker/Linker.tsx:
```diff
@@ -580,7 +615,8 @@ export class Linker extends Component<LinkerProps, LinkerState> {
   isOverlayShown(): boolean {
     const { activeTool } = this.props;
-    return activeTool === ToolType.LINKER;
+    const { isDraggingPanel } = this.state;
```
I didn't realize we don't want the fade transition for this case, so it should be a separate boolean from `isOverlayShown()`.
Just wrap the whole CSSTransition in `!isDraggingPanel ? <CSSTransition in={isOverlayShown} ... > : null`, then you'll have a fade when activating/deactivating the linker by changing the tool, but it'll still be instant when dragging the panel.

Even with these changes, it looks like Linker gets a couple renders in before receiving the `.DRAGGING` event and still logs the error a couple time.  I think logging an error when trying to draw the link may be a bit aggressive. Changing that to `log.warn` instead may be useful: https://github.com/deephaven/web-client-ui/blob/f5794d797403eb3ab2c553d237de11171ac7c9e6/packages/dashboard-core-plugins/src/linker/LinkerOverlayContent.tsx#L291

mofojed commented on packages/dashboard-core-plugins/src/linker/Linker.tsx:
```diff
@@ -510,6 +521,22 @@ export class Linker extends Component<LinkerProps, LinkerState> {
   }
 
+  handlePanelDragging(componentId: string): void {
+    const { links } = this.props;
+    for (let i = 0; i < links.length; i += 1) {
+      const link = links[i];
+      const { start, end } = link;
+      if (start.panelId === componentId || end?.panelId === componentId) {
+        this.setState({ isDraggingPanel: true });
+        return;
+      }
+    }
```
We probably don't even really need a check here, we could just always hide the links when dragging... but whatever, no biggie.
